Separate tmpDir from Windows temp dir & stop running Linux processes on Windows volume

AndrewPardoe opened the pull request:
Merge changes to create winDir in addition to tmpDir.

AndrewPardoe commented:
I'm working slowly towards making the WSL port a bit less of a hack. Previously, I created a user option to set a `tmpDir` through options where `exec.js` does all its work. If you were running under WSL, you'd automatically get a hard-coded `tmpDir` of `c:\tmp`. 

This broke the scenario where people ran CE on WSL without any intention of using MSVC (see issue #468.) If you ran under WSL without a `c:\tmp` directory it would break all `exec`s. I fixed that with a quick hack--a `--wsl` option--with PR #603. 

This PR does three related things:

1. Create a Windows-specific temporary directory, `winTmp`, that is used only when executing binaries that are located on a mounted volume *and* when running under WSL. If you're not running under WSL, or running a binary that's not located on a mounted volume, `exec`s happen in your normal Linux temporary directory. 

2. Set `winTmp` automatically to the value of `%TEMP%` from Windows. The canonicalization of the path can be improved as WSL has just started to ship a path canonicalizer (`/bin/wslpath`) but it's too new to rely on it being available. 

3. Removes the `--wsl` hack which I'm confident is no longer needed. Note that I'm maintaining the `--tmpDir` option in case anyone's using it. It's a nice refinement anyway, I think.

AndrewPardoe commented:
Note that I've tested a few different options on my WSL setup, including no specified temporary directories, specifying a Windows-volume temporary directory, and specifying a Linux-volume temporary directory. (MSVC compilers fail as expected in the last; everything works and looks good in the debugger in the first two.) 

I haven't tested this on a Linux setup at all. I actually thought I was doing a PR from my fork's `tempdir` branch into my fork's `master` branch but GitHub fooled me again. I was actually PRing into the real CE repo.

RabsRincon commented on docs/WindowsSubsystemForLinux.md:
````diff
@@ -0,0 +1,94 @@
+
+The only viable option for debugging under WSL is to use [VS Code](https://code.visualstudio.com). Because VS Code doesn't currently run natively under WSL, you have to attach to a running CE instance. The following is a `launch.json` that works for attaching to an instance of CE that was launched with the `--inspect` flag. 
+
+```json
````
Note that this is not json (No comments in json), and GitHub is complaining. So you either move the comments as a note after the code snippet, or remove the json declaration (and lose the colouring)

AndrewPardoe replied:
I can see that now. I don't think the comments are needed at all (nor is the colorization). I'll fix it to make GitHub happy.

RabsRincon commented on docs/WindowsSubsystemForLinux.md:
```diff
@@ -0,0 +1,94 @@
+- `curl https://raw.githubusercontent.com/creationix/nvm/v0.33.8/install.sh | bash` to install NVM
+- `source ~/.profile` to reload your profile, bringing NVM into your environment
+- `nvm ls-remote -lts` to show the latest supported version of node.js
+- `nvm install 8.9.3`, substituting the latest version, to install node.js
```
It's probably better if you don't suggest a "latest version". We only guarantee that it works on the latest LTS version (Even tho there are probably no reasons for it not to work further down the line)

AndrewPardoe replied:
I'm suggesting you find the LTS on the line above--should I remove all mention of latest, or reword to "latest LTS"?

RabsRincon commented on docs/WindowsSubsystemForLinux.md:
```diff
@@ -0,0 +1,94 @@
+# Running on Windows Subsystem for Linux
+
+Contact: Andrew Pardoe, @AndrewPardoe on GitHub
```
What about `[https://github.com/AndrewPardoe](@AndrewPardoe)` so you get an acutal link? (GitHub should be doing this already, but...)

AndrewPardoe replied:
Yes, will edit that.

RabsRincon commented on lib/base-compiler.js:
```diff
@@ -57,6 +57,13 @@ Compile.prototype.newTempDir = function () {
             else
                 resolve(dirPath);
         });
+        if (process.env.winTmp != process.env.tmpDir && process.env.winTmp != "")
```
Couldn't this be moved to the specific compiler types for wsl? If not, I'd love to know why (I can't seem to find a reason not to, so it would be great to know!) 😄

AndrewPardoe replied:
This could, I think. There are two things that concern me about moving it to WSL-CL.js
1. If ever we have a second (or third!) Windows-only compiler, it'd have to exist in each.
2. This code appears to run once during setup, not during the main edit/compile loop. The code in WSL-CL runs during the loop. I don't know if that's a problem. I'll try moving it.

RabsRincon commented on lib/exec.js:
```diff
@@ -41,13 +41,24 @@ function execute(command, args, options) {
-        detached: process.platform === 'linux'
-    });
+    var child;
+    if (command.startsWith("/mnt") && process.env.wsl) {
```
I think maybe a ternary operator for cwd instead of duplicating the rest of the options is cleaner solution here?

AndrewPardoe replied:
I didn't know you could put a ternary in there :) Will do.

RabsRincon commented on lib/exec.js:
```diff
@@ -41,13 +41,24 @@ function execute(command, args, options) {
+            env: env,
+            detached: process.platform === 'linux'
+        });
+
```
Whitespace

AndrewPardoe replied:
Removed.

RabsRincon commented on app.js:
```diff
@@ -61,16 +61,24 @@ if (opts.debug) logger.level = 'debug';
+    var windowsTemp = child_process.execSync('cmd.exe /c echo %TEMP%').toString().replace(/\\/g, "/");
+    var driveLetter = windowsTemp.substring(0, 1).toLowerCase();
+    var directoryPath = windowsTemp.substring(2).trim();
+    process.env.winTmp = "/mnt/".concat(driveLetter).concat(directoryPath);
```
Just as a minor note, `path.join()` is meant to be used for paths concatenation, but I don't know if it would work for this setup (I'm guessing it does, but you never know!)

AndrewPardoe replied:
Yes, path.join should work here. I don't know that path.split would work as it's running on Linux and getting back a Windows path. Let me look at this.

jaredwy replied:
If you want to do win32 path operations and be sure you are getting that result back use path.win32

https://nodejs.org/api/path.html#path_path_win32

jaredwy replied:
This would be preferable to string operations.

AndrewPardoe replied:
@jaredwy, thank you for the link, but I don't see how there are path.win32 operations that will do this for me. There's a path.win32 function that will return a drive letter as `C:\` but I'd still have to `substring` and `tolower` it in order to turn it into `/mnt/c/`.

RabsRincon commented:
Hello Andrew! Thank you so much for this!
First things first, you don't need to worry about code coverage for WSL-specific code (We use the bot more as a reminder to write tests)
Also, you probably want to edit the repo Wiki to add a link to the newly added doc, so it gets a bit more exposure. Once the small changes I've mentioned are merged, I'll test the PR and if no problems arise, will merge.

Thank you again for the time you're taking to improve WSL support on CE.

AndrewPardoe commented:
@RabsRincon, I think I've addressed all the feedback. But I think I've done one thing wrong. 

I don't understand how `WSL-CL.js` overrides functionality in `base-compiler.js`. Using `logger.info` to trace, it seems as if the WSL version of `Compile.prototype.newTempDir` is being called all the time. Is this function supposed to have a different name much like `CompileCl` and its siblings for Go, Haskell, etc.? What's the mechanism by which this overriding is performed?

RabsRincon commented on lib/compilers/WSL-CL.js:
```diff
@@ -64,4 +66,17 @@ function compileCl(info, env) {
 
+// AP: Create CE temp directory in winTmp directory. This code matches overrides the function in base-compiler.js. 
+// NPM temp package: https://www.npmjs.com/package/temp, see Affixes
+Compile.prototype.newTempDir = function () {
```
So the override mechanism works like this:
The compiler specifies a type on the properties, which is used to load a `compiler/TYPE.js` file, and we get whatever it outputs as a kind of "wrapper" around the compiler.
Now, insde the function that gets exported, you initialize a new Compile object, in which you change the objects that need to be changed, but only to that Compile object, not the global one (As you're doing right now)

So you need to move this `Compiler.prototype.newTempDir` function inside the compileCl function (after L37 `var compile ...` but before the initialization at L66) and overrride the newTempDir like `compile.newTempDir = function () { ... };`
I hope some of this made any sense

AndrewPardoe replied:
Yes, that makes sense. I still need to read the code more carefully, but I've pushed the changes. And my printf debugging says you're correct as well :)
